### DISCUSSION TITLE
서브넷 마스크란?

### DISCUSSION THREAD

**Discussion by seunghye218** [2024-woowacourse-study/level-interview#212] [Network]
.

**Comment by nayonsoso** [comment]
### IP의 구성
IP는 4Byte = 32bit 로 구성되어있다.
IP 는 한 바이트씩 점을 찍어 표현하는데, (e.g. 127.0.0.1) 여기서 점으로 나누어진 각 부분을 `옥텟`이라고 한다.
IP 의 32bit는 `Network ID` 와 `Host ID`로 이루어져있다.
Network ID는 어떤 네트워크인지 식별하기 위해 사용되고,
Host ID는 그 네트워크 안에서 어떤 컴퓨터인지 식별하기 위해 사용된다.

### 서브넷 마스크란
- 서브넷 마스크를 통해서 네트워크를 서브넷으로 나눌 수 있다.
- 서브넷 마스크는 32개의 연속된 1과 연속된 0으로 이루어져있다.
    - IP와 비슷하다고 생각할 수 있지만, 전혀 다른 개념이니 주의!
- 여기서 연속된 1이 Network ID를, 연속된 0이 Host ID를 나타낸다.
- 따라서 서브넷 마스크와 IP 주소를 AND 연산한다면, Network ID를 알 수 있다. (이런 점에서 ‘마스크’라고 하는 듯)

### 구식 IP 할당 방법 - Class Addressing
- Network ID와 Host ID의 경계(서브넷 마스크)를 옥텟을 기준으로 나누어 클래스로 구분하는 방식이다.
  - A클래스 : 첫번째 옥텟까지 Network ID
  - B클래스 : 두번째 옥텟까지 Network ID
  - C클래스 : 세번째 옥텟까지 Network ID
- IPv4는 초기에는 부족한 IP 주소를 해결하기 위해 이 방식을 사용했다.
- 하지만 옥텟 단위로만 Network ID 와 Host ID를 나누는 것은 비효율적이었으며, 더 적절한 크기로 네트워크를 분할할 필요성이 생기게 되었다.
- 따라서 클래스로 네트워크를 나누는 방식은 지금은 더 이상 사용되지 않고, 다른 방식(CIDR 방식)으로 할당하도록 1993년도에 바뀌었다.

### 최신식 IP 할당 방법 - CIDR
- 직접 서브넷 마스크를 설정하는 방법이다.
- CIDR에서는 서브넷 마스크에서 IP에서 몇개의 연속된 1이 있는지, 즉 네트워크 ID가 몇 bit 까지인지를 간단히 나타낼 수 있다.
- IP 뒤에 `/n` 을 붙여 표현한다.
- 예를 들어, IP 주소가 172.30.90.149 이고, 서브넷 마스크가 255.240.0.0 이라면 `172.30.90.149/12` 라고 나타낼 수 있다.

**Comment by gyungchan-jo** [comment]
서브넷 마스크는 1과 0으로 이루어진 32비트 숫자입니다. 1은 네트워크 부분을, 0은 호스트 부분을 나타냅니다. 
- 네트워크 부분: 네트워크를 식별하는 데 사용됩니다.
- 호스트 부분: 네트워크 내의 개별 장치를 식별하는 데 사용됩니다.

예를 들어, 서브넷 마스크 255.255.255.0은 다음과 같이 표현됩니다:
이진수: 11111111.11111111.11111111.00000000

📌 서브넷 마스크의 예시
IP 주소: 192.168.1.10
서브넷 마스크: 255.255.255.0

이 경우, 서브넷 마스크는 다음과 같이 해석됩니다:
네트워크 부분: 192.168.1
호스트 부분: 10

이러한 서브넷 마스크(Subnet Mask)는 IP 주소를 네트워크 부분과 호스트 부분으로 나누기 위해 사용되는 32비트 숫자입니다. 이는 네트워크를 더 작은 서브네트워크(서브넷)로 나누는 데 사용되며, 네트워크 관리와 효율성을 높이는 데 중요한 역할을 합니다.
